### PR TITLE
Added support for collections of Parceler values

### PR DESCRIPTION
**Pull request by iainconnor**
Fixing [#72](https://github.com/f2prateek/dart/issues/72).

### CODE REVIEW

**Comment by stephanenicolas** [comment]
@f2prateek can you take care of this please ?

The PR looks great to me. S.

2015-10-16 12:28 GMT-07:00 Iain Connor <notifications@github.com>:

**Comment by stephanenicolas** [comment]
The PR looks good. Can you just add 1 or 2 tests please ? @iainconnor

**Comment by iainconnor** [comment]
Sure, where would you like to see the tests? Do they make sense in the Dart module or the processor?(Or the Henson processor?)

**Comment by stephanenicolas** [comment]
test that you can pass lists and sparse array as requested. There are already some tests for that (I don't remember where). A priori in Dart, not in Henson.

**Comment by f2prateek** [comment]
@iainconnor they tests would go here - https://github.com/f2prateek/dart/blob/master/dart-processor/src/test/java/com/f2prateek/dart/processor/InjectExtraTest.java.

Essentially write a sample of the code, and verify that the generated source code matched what was expected.

**Comment by stephanenicolas** [comment]
Shouldn't we also test that the collection bundle and unbundles fine as
well ?
Here :
https://github.com/f2prateek/dart/blob/master/dart-sample/src/test/java/com/f2prateek/dart/example/SampleActivityTest.java

That would also give an example.

2015-10-16 16:00 GMT-07:00 Prateek Srivastava <notifications@github.com>:

**Comment by johncarl81** [comment]
@iainconnor, I know this is a rare case, but you could have a Parceler collection that is deeper than one level:
```
@Parcel class Example{}
List<List<Example>>
Map<String, List<Example>>
```

What do you think about recursively searching the generics tree for `@Parcel`?

**Comment by stephanenicolas** [comment]
Interesting edge case. The recursive approach would solve this problem once
for all...

**Comment by stephanenicolas** [comment]
@johncarl81 @iainconnor that's our week before we release RC 2. Does anyone want to fix the comments ?

**Comment by iainconnor** [comment]
What do we feel about the performance of a recursive traversal?

**Comment by johncarl81** [comment]
The performance should be negligible, especially during compile time.  This could also be done iteratively.

**Comment by johncarl81** [comment]
FYI: https://github.com/iainconnor/dart/pull/1

**Comment by stephanenicolas** [comment]
Did you do a PR against dart ?

S.

2016-02-22 22:08 GMT-08:00 John Ericksen <notifications@github.com>:

**Comment by iainconnor** [comment]
Thanks @johncarl81. @stephanenicolas just have to resolve some conflicts.

**Comment by johncarl81** [comment]
I'd consider a rebase @iainconnor.

File: dart-common/src/main/java/com/f2prateek/dart/common/AbstractDartProcessor.java [hunk 1]
```diff
@@ -283,6 +289,72 @@ private void parseInjectExtra(Element element, Map<TypeElement, InjectionTarget>
...
 
+  private boolean isParcel(TypeMirror type) {
+    List<Element> supportedTypes = getTypeElements(
+            new Class[]{Byte.class, Double.class, Float.class, Integer.class,
+                        Long.class, Character.class, Boolean.class, String.class});
+
+    List<Element> singleGenericCollections = getTypeElements(
+            new Class[]{List.class, ArrayList.class, LinkedList.class,
+                        Set.class, HashSet.class, SortedSet.class, TreeSet.class,
+                        LinkedHashSet.class});
```
Comment (dlemures) [inline dart-common/src/main/java/com/f2prateek/dart/common/AbstractDartProcessor.java]: What would happen with SparseArray?

Comment (johncarl81) [inline dart-common/src/main/java/com/f2prateek/dart/common/AbstractDartProcessor.java]: Nothing at the moment... you could add that to the `supportedTypes` as it's not a generic collection.  There are a few other misc types I didn't map like `Bundle` or `IBinder`.  Here's the list Parceler supports:  https://github.com/johncarl81/parceler/blob/master/parceler/src/main/java/org/parceler/internal/ParcelerModule.java#L219-L258

**Comment by dlemures** [comment]
:+1:

**Comment by johncarl81** [comment]
I'm considering moving this code into a shared utility library within the
parceler codebase as there are a few other implementations very similar to
this.  Would you like this utility for the upcoming release?

File: dart-common/src/main/java/com/f2prateek/dart/common/AbstractDartProcessor.java (hunk 1 above)
Comment (stephanenicolas) [inline dart-common/src/main/java/com/f2prateek/dart/common/AbstractDartProcessor.java]: Seems to me that maps and hashmaps, bundle, sparse aarray should all go in

Comment (johncarl81) [inline dart-common/src/main/java/com/f2prateek/dart/common/AbstractDartProcessor.java]: They should be easy to add.

File: dart-common/src/main/java/com/f2prateek/dart/common/AbstractDartProcessor.java [hunk 2]
```diff
@@ -283,6 +289,72 @@ private void parseInjectExtra(Element element, Map<TypeElement, InjectionTarget>
...
+
+  private boolean isParcelRecursive(TypeMirror type,
+                                    List<Element> supportedTypes,
+                                    List<Element> singleGenericCollections,
+                                    List<Element> doubleGenericCollections,
+                                    boolean subCollection) {
+    if (subCollection && existsWithin(type, supportedTypes)) {
+      return true;
+    }
+    if (hasAnnotationWithFqcn(typeUtils.asElement(type), "org.parceler.Parcel")) {
```
Comment (stephanenicolas) [inline dart-common/src/main/java/com/f2prateek/dart/common/AbstractDartProcessor.java]: Using Parceler.class.getName() is more robust.

**Comment by dlemures** [comment]
@johncarl81 I think it would be nice to have it.
We can merge this PR now and update to use the utility when it is ready

File: dart-common/src/main/java/com/f2prateek/dart/common/AbstractDartProcessor.java (hunk 2 above)
Comment (johncarl81) [inline dart-common/src/main/java/com/f2prateek/dart/common/AbstractDartProcessor.java]: That does require that Parceler is on the classpath, the current technique allows for Parceler to be optional.

**Pull request merged**